libqrexec/exec.c

```diff
@@ -105,4 +136,240 @@ void do_fork_exec(const char *cmdline, int *pid, int *stdin_fd, int *stdout_fd,
+            }
+            memcpy(service_full_path, qubes_rpc_directories[i].string, directory_length);
+            memcpy(service_full_path + directory_length, command->service_descriptor, total_path_length - directory_length);
+            service_full_path[total_path_length] = '\0';
+
+            if (!qubes_connect(s, service_full_path, total_path_length)) {
+                *stdout_fd = *stdin_fd = s;
+                *stderr_fd = -1;
+                *pid = -1;
+                set_nonblock(s);
+                return 0;
+            }
+            switch (errno) {
+            // These cannot happen
+            case EFAULT:       // all of our parameters are in valid memory
+            case EINVAL:       // we passed valid parameters
+            case ENOTSOCK:     // ditto
+            case ENETUNREACH:  // cannot happen for AF_UNIX
+            case EADDRNOTAVAIL:// cannot happen for AF_UNIX
+            case EINPROGRESS:  // this socket is blocking
+            case EAGAIN:       // ditto
+#if EAGAIN != EWOULDBLOCK
+            case EWOULDBLOCK:  // ditto
+#endif
+            case EINTR:        // we already check for this and retry
+            case EBADF:        // `s` was created by a call to `socket()`
+            case EAFNOSUPPORT: // the kernel supports AF_UNIX
+            case EALREADY:     // ditto
+            default:
+                perror("unexpected return from connect()");
+                abort();
+            // Filesystem misconfigured
+            case ENOENT:
+            case ENOTDIR:
+                // These errors would also happen with `execve()`
+                break;
+            // These should not happen
+            case EPROTOTYPE:
+            case ETIMEDOUT:
+            case EPERM:
+            case EIO:
+            case EISDIR:
+                // Socket server problem and/or misconfiguration.  Fail the whole connection.
+                // (We do not want to fall back because the user may have
+                // overriden behavior for security reasons, and because
+                // fail-fast is much easier to debug).
+                goto out;
+            // Found a file instead of a socket
+            case EACCES:
+            case ECONNREFUSED: {
+                close(s);
+                const size_t remote_domain_length = strlen(command->remote_domain);
+                // cannot overflow due to command length limits
+                size_t username_length = 0, exec_cmdline_length =
+                    command->service_descriptor_length + remote_domain_length + 2;
+                if (command->username) {
+                    username_length = strlen(command->username);
+                    exec_cmdline_length += username_length + 1;
+                }
+                char *const exec_cmdline = calloc(exec_cmdline_length, 1);
+                if (!exec_cmdline) {
+                    fputs("Out of memory!\n", stderr);
+                    abort();
+                } else {
```

Comment:
abort should never return. So you can move code from the else to the normal flow.